runtime/engine/src/test/java/io/aklivity/zilla/runtime/engine/test/internal/validator/TestValidator.java

```diff
@@ -22,28 +22,20 @@
 public class TestValidator implements Validator
 {
     @Override
-    public boolean read(
+    public DirectBuffer read(
         DirectBuffer data,
         int index,
         int length)
     {
-        return validate(data, index, length);
+        return length == 18 ? data : null;
```

Comment:
This is not quite right because it loses information about index and length.

Take a look at `MessageConsumer`, `MessageFunction` and `MessagePredicate` in `io.aklivity.zilla.runtime.engine.binding.function` package.

I think we might need the following in `io.aklivity.zilla.runtime.engine.validator.function` package:
 - `ToIntFragmentFunction`

```
ToIntFragmentFunction
  int applyAsInt(int flags, DirectBuffer buffer, int index, int length)
```

```
Validator
  int validate(int flags, DirectBuffer buffer, int index, int length, ToIntFragmentFunction next)
```

The return value from `applyAsInt` is the progress towards `length`, and the implementation would write the validated (converted) value to `next`.

```
public class TestValidator implements Validator
{
    public DirectBuffer validate(
        DirectBuffer data,
        int index,
        int length,
        ToIntFragmentFunction next)
    {
        return length != 18 ? -1 : next.applyAsInt(flags, buffer, index, length);
    }
}
```

So the lambda we pass in for `next` would actually write the value into the next step in the pipeline.

A better implementation would track the state of progress towards 18 bytes in this case, but that makes it stateful, which is another level of complexity we need to sort out after getting the interfaces to support streaming.

What do you think?

Reply:
Alternatively, we can consider using `MessageConsumer` more directly in validator, if that makes it simpler to propagate valid `DATA` frames.

```
public class TestValidator implements Validator
{
    public int validate(
        int msgTypeId, // DataFW.TYPE_ID
        DirectBuffer data,
        int index,
        int length,
        MessageConsumer next)
    {
        return length != 18 ? -1 : next.applyAsInt(msgTypeId, buffer, index, length);
    }
}
```